Remotion/Data/DomainObjects/Persistence/Rdbms/DbCommandBuilders/QueryDbCommandBuilder.cs

```diff
@@ -48,23 +48,24 @@ public override IDbCommand Create (IRdbmsProviderCommandExecutionContext command
       var command = commandExecutionContext.CreateDbCommand();
 
       var statement = _statement;
-      foreach (var parameterWithType in _parametersWithType)
+      foreach (var parameterWithDefinition in _parametersWithDefinition)
       {
-        var queryParameterValue = parameterWithType.QueryParameter.Value;
+        var queryParameterValue = parameterWithDefinition.QueryParameter.Value;
 
-        if (parameterWithType.QueryParameter.ParameterType == QueryParameterType.Text)
+        if (parameterWithDefinition.QueryParameter.ParameterType == QueryParameterType.Text)
         {
           Assertion.DebugAssert(
               queryParameterValue is string,
               "parameterWithType.QueryParameter.Value is string when parameterWithType.QueryParameter.ParameterType == Text");
           var queryParameterValueAsString = (string)queryParameterValue;
 
-          statement = statement.Replace(parameterWithType.QueryParameter.Name, queryParameterValueAsString);
+          statement = statement.Replace(parameterWithDefinition.QueryParameter.Name, queryParameterValueAsString);
         }
         else
         {
-          var parameter = parameterWithType.StorageTypeInformation.CreateDataParameter(command, queryParameterValue);
-          parameter.ParameterName = parameterWithType.QueryParameter.Name;
+          var dataParameterValue = parameterWithDefinition.DataParameterDefinition.GetParameterValue(queryParameterValue);
+          var parameter = parameterWithDefinition.DataParameterDefinition.CreateDataParameter(command, dataParameterValue);
+          parameter.ParameterName = parameterWithDefinition.QueryParameter.Name;
```

Comment:
it feels weird that CreateDataParameter isn't setting the name since it already has all the relevant information. please check if this should be changed. One reason that might argue against it would be that we have to duplicate the setting of the name-property into all implementations. since we do need to set value etc, too, might be with it. 

Another reason against it might be that it clashes when we do structured data or collection parameters.

If we don't do it internally, I wonder if this might warrent an inline comment to explain why we don't do it since it does feel like it should be handled inside CreateDataParameter.

Reply:
It makes sense to do it within `IDataParameterDefinition.CreateDataParameter()`, so I implemented it in all 3 basic implementations.
It will not clash with structured data types or collections, because it is still a single parameter with one name; we just have to implement setting the name a fourth time.

Reply:
```
var dataParameterValue = parameterWithDefinition.DataParameterDefinition.GetParameterValue(queryParameterValue);
var parameter = parameterWithDefinition.DataParameterDefinition.CreateDataParameter(command, dataParameterValue);
parameter.ParameterName = parameterWithDefinition.QueryParameter.Name;
```
I realized a misunderstanding on my part: I read the reviewed code as the DataParameterDefinition already owning the QueryParameter object and thus the name. Which isn't the case. no we have a new method argument and the duplication. I'm okay with that if you say it's cleaner / more helpful than what we had before. Am just mentioning it because it's a review-mistake.

Am commenting/unresolving this while checking the rest. Assuming there won't be any other details I'll approve it as is and leave the final decision on the dataParameterName up to you.